events.go

```diff
@@ -68,6 +68,13 @@ type ConnectReply struct {
 	// QueueInitialCap set an initial capacity for client's message queue, the size of queue
 	// can grow further, but won't be reduced below QueueInitialCap. By default, it's 2.
 	QueueInitialCap int
+	// ClientPingInterval sets the ping interval for each client. Note that to
```

Comment:
Thanks a lot for the pr @bfwbbrj!

The behaviour described here seems not very obvious to me. And it looks like this is a consequence of the complexity of current ping/pong configuration in the library.

One possible approach here could be adding `AppLevelPing *AppLevelPing` to this struct instead of `ClientPingInterval` and `ClientPongTimeout`. If not nil - it could override the ping/pong configuration coming from the connection transport level. I think this may be the most straightforward way to override ping/pong with the current Centrifuge API.

Though probably we can do better. The complexity of current solution partly comes from the fact we had to support both ClientProtocolV1 and ClientProtocolV2. But in https://github.com/centrifugal/centrifuge/pull/284 we are removing ClientProtocolV1.

In general, the idea in ping/pong configuration was to achieve the following:

* possibility to configure ping/pong on transport level
* sane defaults used if nothing configured (most common scenario), where we have 25/10 secs.
* way to override values for all transport connections (valid scenario, as you mentioned different apps may have different requirements)
* possibility to disable ping/pong using -1 (rare scenario)

Now since we need to override on a client level to support multi-tenancy – we should override values coming from the connection transport, and keep all the properties above. So the alternative way which I see here may be removing `AppLevelPing` from Centrifuge entirely, and only leave `PingPongConfig`. `Transport` should return the configured `PingPongConfig` without any change to the `Client`. I.e. have `PingPongConfig()` method instead of `AppLevelPing()`. Then, in the `ConnectReply`, we can have `PingPong *PingPongConfig` to override values coming from the `Transport`. I think this will make the configuration more obvious and only one Ping/Pong-related struct will exist in Centrifuge.

I'd go with second suggestion in terms of v0.29.x. Any thought? Probably you see even better configuration approach.

Reply:
@FZambia Totally agreed. I want to double check that [transport level ping/pong](https://github.com/bfwbbrj/centrifuge/blob/ccf897517a1838a22f89d50fb1fa104befb29f5c/handler_websocket.go#L344) will be removed after #284 as you mentioned, right? Otherwise there will be duplicate ping/pong on the transport level. And can I amend the current commit or do I need to create a new commit for this?

I also have some unrelated questions. Why are we calling [stopTimer](https://github.com/bfwbbrj/centrifuge/blob/ccf897517a1838a22f89d50fb1fa104befb29f5c/client.go#L443) and creating new [Timers](https://github.com/bfwbbrj/centrifuge/blob/ccf897517a1838a22f89d50fb1fa104befb29f5c/client.go#L470) here? Imho, we can reuse the same timer for each client to save the memory&cpu usage.

Reply:
> I want to double check that [transport level ping/pong](https://github.com/bfwbbrj/centrifuge/blob/ccf897517a1838a22f89d50fb1fa104befb29f5c/handler_websocket.go#L344) will be removed after https://github.com/centrifugal/centrifuge/pull/284 as you mentioned, right?

For bidirectional websocket yes - most probably we won't have code which sends control websocket PING frames in v0.29.0.

> And can I amend the current commit or do I need to create a new commit for this?

In general, if you open a completely different PR - it will be simpler for me. So if you want to try implementing suggestion #2 from comment above - open it over v0.29.x branch.

> I also have some unrelated questions. Why are we calling [stopTimer](https://github.com/bfwbbrj/centrifuge/blob/ccf897517a1838a22f89d50fb1fa104befb29f5c/client.go#L443) and creating new [Timers](https://github.com/bfwbbrj/centrifuge/blob/ccf897517a1838a22f89d50fb1fa104befb29f5c/client.go#L470) here? Imho, we can reuse the same timer for each client to save the memory&cpu usage.

Currently we are using one timer for each tasks of connection. I think your idea makes sense in general - but it can bring its own issues, must be very carefully implemented, with making sure we always have timer to execute connection operation, and no queuing happens. Also benchmarks which prove this makes sense, etc. My thinking here is that optimization like this should araise from real app profiles (where we usually see syscalls now). If you are talking about simply using pool of timers - then I am not sure it makes sense here as connections are usually long-term.